### ISSUE TITLE
VW Streaming Job get stuck.

### ISSUE THREAD

**Issue by kemaswill** [JohnLangford/vowpal_wabbit#1569]
Run vw in a streaming way: `hadoop jar hadoop-streaming.jar -D mapreduce.map.memory.mb=5120 -D mapreduce.map.java.opts=-Xmx3584m  -D mapreduce.job.acl-view-job=* -D mapred.job.queue.name=$queueName -input $modelIn -output $modelOut -mapper "runvw.sh ${modelOut} $host 24 25 logistic" -file vw -file runvw.sh 2>&1`

The main content of runvw.sh is `./vw -b $bit_precision --total $mapreduce_job_maps --node $mapper --unique_id $mapred_job_id --passes $passes -c --span_server $spanning_tree_host --readable_model readable_model --loss_function=$loss_function $cmd_line_args`

However, the mapreduce job get stuck after processing 67% percentage, following is the errors, don't know whether it's related to this [issue](https://github.com/JohnLangford/vowpal_wabbit/issues/1039), and wonder what does 'ou need to trick Hadoop into having a fixed-size set of mappers because
synchronous communication is required. To do this partition the dataset
with .gz files (which Hadoop cannot break up) or run VW in a (controlled)
number of reducers.' mean? Is there any doc on this?

> 18/08/15 05:19:47 INFO mapreduce.Job:  map 64% reduce 0%
18/08/15 05:19:50 INFO mapreduce.Job:  map 65% reduce 0%
18/08/15 05:19:52 INFO mapreduce.Job:  map 66% reduce 0%
18/08/15 05:20:00 INFO mapreduce.Job:  map 67% reduce 0%
18/08/15 05:21:20 INFO mapreduce.Job: Task Id : attempt_1534162772717_315512_m_000811_0, Status : FAILED
Error: java.lang.RuntimeException: PipeMapRed.waitOutputThreads(): subprocess failed with code 141
    at org.apache.hadoop.streaming.PipeMapRed.waitOutputThreads(PipeMapRed.java:322)
    at org.apache.hadoop.streaming.PipeMapRed.mapRedFinished(PipeMapRed.java:535)
    at org.apache.hadoop.streaming.PipeMapper.close(PipeMapper.java:130)
    at org.apache.hadoop.mapred.MapRunner.run(MapRunner.java:61)
    at org.apache.hadoop.streaming.PipeMapRunner.run(PipeMapRunner.java:34)
    at org.apache.hadoop.mapred.MapTask.runOldMapper(MapTask.java:453)
    at org.apache.hadoop.mapred.MapTask.run(MapTask.java:343)
    at org.apache.hadoop.mapred.YarnChild$2.run(YarnChild.java:175)
    at java.security.AccessController.doPrivileged(Native Method)
    at javax.security.auth.Subject.doAs(Subject.java:422)
    at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1953)
    at org.apache.hadoop.mapred.YarnChild.main(YarnChild.java:169)

18/08/15 05:21:21 INFO mapreduce.Job: Task Id : attempt_1534162772717_315512_m_001241_0, Status : FAILED
Error: java.lang.RuntimeException: PipeMapRed.waitOutputThreads(): subprocess failed with code 141
    at org.apache.hadoop.streaming.PipeMapRed.waitOutputThreads(PipeMapRed.java:322)
    at org.apache.hadoop.streaming.PipeMapRed.mapRedFinished(PipeMapRed.java:535)
    at org.apache.hadoop.streaming.PipeMapper.close(PipeMapper.java:130)
    at org.apache.hadoop.mapred.MapRunner.run(MapRunner.java:61)
    at org.apache.hadoop.streaming.PipeMapRunner.run(PipeMapRunner.java:34)
    at org.apache.hadoop.mapred.MapTask.runOldMapper(MapTask.java:453)
    at org.apache.hadoop.mapred.MapTask.run(MapTask.java:343)
    at org.apache.hadoop.mapred.YarnChild$2.run(YarnChild.java:175)
    at java.security.AccessController.doPrivileged(Native Method)
    at javax.security.auth.Subject.doAs(Subject.java:422)
    at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1953)
    at org.apache.hadoop.mapred.YarnChild.main(YarnChild.java:169)

...
...
...

18/08/15 05:31:05 INFO mapreduce.Job: Task Id : attempt_1534162772717_315512_m_001196_0, Status : FAILED
AttemptID:attempt_1534162772717_315512_m_001196_0 Timed out after 600 secs
18/08/15 05:31:05 INFO mapreduce.Job: Task Id : attempt_1534162772717_315512_m_001601_0, Status : FAILED
AttemptID:attempt_1534162772717_315512_m_001601_0 Timed out after 600 secs
18/08/15 05:31:05 INFO mapreduce.Job: Task Id : attempt_1534162772717_315512_m_001115_0, Status : FAILED
AttemptID:attempt_1534162772717_315512_m_001115_0 Timed out after 600 secs
18/08/15 05:31:05 INFO mapreduce.Job: Task Id : attempt_1534162772717_315512_m_001439_0, Status : FAILED
AttemptID:attempt_1534162772717_315512_m_001439_0 Timed out after 600 secs
18/08/15 05:31:05 INFO mapreduce.Job: Task Id : attempt_1534162772717_315512_m_000058_0, Status : FAILED
AttemptID:attempt_1534162772717_315512_m_000058_0 Timed out after 600 secs
18/08/15 05:31:05 INFO mapreduce.Job: Task Id : attempt_1534162772717_315512_m_001277_0, Status : FAILED
AttemptID:attempt_1534162772717_315512_m_001277_0 Timed out after 600 secs
18/08/15 05:31:05 INFO mapreduce.Job: Task Id : attempt_1534162772717_315512_m_001456_0, Status : FAILED
AttemptID:attempt_1534162772717_315512_m_001456_0 Timed out after 600 secs
18/08/15 05:31:05 INFO mapreduce.Job: Task Id : attempt_1534162772717_315512_m_001358_0, Status : FAILED

...
...
...

18/08/15 05:33:51 INFO mapreduce.Job: Task Id : attempt_1534162772717_315512_m_000091_1, Status : FAILED
AttemptID:attempt_1534162772717_315512_m_000091_1 Timed out after 600 secs
Sent signal OUTPUT_THREAD_DUMP (SIGQUIT) to pid 44714 as user jwpan for container container_e09_1534162772717_315512_01_001815, result=success
Container killed by the ApplicationMaster.
Container killed on request. Exit code is 143
Container exited with a non-zero exit code 143

18/08/15 05:33:52 INFO mapreduce.Job:  map 30% reduce 0%
18/08/15 05:33:54 INFO mapreduce.Job:  map 31% reduce 0%
18/08/15 05:33:59 INFO mapreduce.Job:  map 32% reduce 0%
18/08/15 05:34:02 INFO mapreduce.Job:  map 33% reduce 0%
18/08/15 05:34:05 INFO mapreduce.Job: Task Id : attempt_1534162772717_315512_m_001124_1, Status : FAILED
AttemptID:attempt_1534162772717_315512_m_001124_1 Timed out after 600 secs
18/08/15 05:34:06 INFO mapreduce.Job:  map 34% reduce 0%
18/08/15 05:34:11 INFO mapreduce.Job:  map 35% reduce 0%
18/08/15 05:34:17 INFO mapreduce.Job:  map 36% reduce 0%
18/08/15 05:34:27 INFO mapreduce.Job:  map 37% reduce 0%
18/08/15 05:34:37 INFO mapreduce.Job:  map 38% reduce 0%
18/08/15 05:34:49 INFO mapreduce.Job:  map 39% reduce 0%
18/08/15 05:35:10 INFO mapreduce.Job:  map 40% reduce 0%
18/08/15 05:35:32 INFO mapreduce.Job:  map 41% reduce 0%
18/08/15 05:36:44 INFO mapreduce.Job: Task Id : attempt_1534162772717_315512_m_001270_1, Status : FAILED
Error: java.lang.RuntimeException: PipeMapRed.waitOutputThreads(): subprocess failed with code 143
    at org.apache.hadoop.streaming.PipeMapRed.waitOutputThreads(PipeMapRed.java:322)
    at org.apache.hadoop.streaming.PipeMapRed.mapRedFinished(PipeMapRed.java:535)
    at org.apache.hadoop.streaming.PipeMapper.close(PipeMapper.java:130)
    at org.apache.hadoop.mapred.MapRunner.run(MapRunner.java:61)
    at org.apache.hadoop.streaming.PipeMapRunner.run(PipeMapRunner.java:34)
    at org.apache.hadoop.mapred.MapTask.runOldMapper(MapTask.java:453)
    at org.apache.hadoop.mapred.MapTask.run(MapTask.java:343)
    at org.apache.hadoop.mapred.YarnChild$2.run(YarnChild.java:175)
    at java.security.AccessController.doPrivileged(Native Method)
    at javax.security.auth.Subject.doAs(Subject.java:422)
    at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1953)
    at org.apache.hadoop.mapred.YarnChild.main(YarnChild.java:169)

**Comment by kemaswill** [comment]
Here is the error info from stderr: 

> average    since         example     example  current  current  current
> loss       last          counter      weight    label  predict features
> 0.693147   0.693147            1         1.0  -1.0000   0.0000       21
> 0.497471   0.301794            2         2.0  -1.0000  -1.0433       22
> 0.355580   0.213689            4         4.0  -1.0000  -0.6292       20
> 0.313269   0.270957            8         8.0  -1.0000  -1.5795       22
> 3.518362   6.723455           16        16.0  -1.0000  -1.2051       20
> 2.044455   0.570549           32        32.0  -1.0000  -0.2724       20
> 1.213148   0.381842           64        64.0  -1.0000  -2.1385       22
> 0.750227   0.287306          128       128.0  -1.0000  -0.7461       15
> 0.486924   0.223620          256       256.0  -1.0000  -1.7846       21
> 0.396786   0.306649          512       512.0  -1.0000  -2.2938       20
> 0.346758   0.296730         1024      1024.0  -1.0000  -4.4544       20
> 0.308301   0.269844         2048      2048.0  -1.0000  -3.0043       21
> 0.304058   0.299815         4096      4096.0  -1.0000  -2.1835       22
> 0.291618   0.279179         8192      8192.0  -1.0000  -1.0077       22
> connecting to 10.211.222.85 = gwbl825n01.blue.ygrid.yahoo.com:26543
> connect attempt 1 failed: Connection refused
> connect attempt 2 failed: Connection refused
> connect attempt 3 failed: Connection refused
> connect attempt 4 failed: Connection refused
> connect attempt 5 failed: Connection refused
> connect attempt 6 failed: Connection refused
> connect attempt 7 failed: Connection refused
> connect attempt 8 failed: Connection refused
> connect attempt 9 failed: Connection refused
> connect attempt 10 failed: Connection refused
> connect attempt 11 failed: Connection refused
> connect attempt 12 failed: Connection refused
> connect attempt 13 failed: Connection refused
> connect attempt 14 failed: Connection refused
> connect attempt 15 failed: Connection refused
> connect attempt 16 failed: Connection refused

After attempts for 100 times, the job failed:

> vw: std::exception

The error from stdout is:

> 2018-08-15 15:53:45,873 INFO [main] org.apache.hadoop.streaming.PipeMapRed: PipeMapRed failed!
> java.lang.RuntimeException: PipeMapRed.waitOutputThreads(): subprocess failed with code 1
> 	at org.apache.hadoop.streaming.PipeMapRed.waitOutputThreads(PipeMapRed.java:322)
> 	at org.apache.hadoop.streaming.PipeMapRed.mapRedFinished(PipeMapRed.java:535)
> 	at org.apache.hadoop.streaming.PipeMapper.close(PipeMapper.java:130)
> 	at org.apache.hadoop.mapred.MapRunner.run(MapRunner.java:61)
> 	at org.apache.hadoop.streaming.PipeMapRunner.run(PipeMapRunner.java:34)
> 	at org.apache.hadoop.mapred.MapTask.runOldMapper(MapTask.java:453)
> 	at org.apache.hadoop.mapred.MapTask.run(MapTask.java:343)
> 	at org.apache.hadoop.mapred.YarnChild$2.run(YarnChild.java:175)
> 	at java.security.AccessController.doPrivileged(Native Method)
> 	at javax.security.auth.Subject.doAs(Subject.java:422)
> 	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1953)
> 	at org.apache.hadoop.mapred.YarnChild.main(YarnChild.java:169)
> 2018-08-15 15:53:45,879 WARN [main] org.apache.hadoop.mapred.YarnChild: Exception running child : java.lang.RuntimeException: PipeMapRed.waitOutputThreads(): subprocess failed with code 1
> 	at org.apache.hadoop.streaming.PipeMapRed.waitOutputThreads(PipeMapRed.java:322)
> 	at org.apache.hadoop.streaming.PipeMapRed.mapRedFinished(PipeMapRed.java:535)
> 	at org.apache.hadoop.streaming.PipeMapper.close(PipeMapper.java:130)
> 	at org.apache.hadoop.mapred.MapRunner.run(MapRunner.java:61)
> 	at org.apache.hadoop.streaming.PipeMapRunner.run(PipeMapRunner.java:34)
> 	at org.apache.hadoop.mapred.MapTask.runOldMapper(MapTask.java:453)
> 	at org.apache.hadoop.mapred.MapTask.run(MapTask.java:343)
> 	at org.apache.hadoop.mapred.YarnChild$2.run(YarnChild.java:175)
> 	at java.security.AccessController.doPrivileged(Native Method)
> 	at javax.security.auth.Subject.doAs(Subject.java:422)
> 	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1953)
> 	at org.apache.hadoop.mapred.YarnChild.main(YarnChild.java:169)

@JohnLangford Any thoughts? Should I try to control the number of mappers as you suggested?

**Comment by kemaswill** [comment]
Problem resolved, the spanning_tree process is not started correctly.

**Comment by arielf** [comment]
The info below is probably no longer helpful, given the additional information given later
(java stack trace from the mapreduce)  nevertheless, it is useful to know that:

Unix exit status has useful info in it:

The original failure exit code is `141`.  Unix exit status is 8 bits (so more info can be added os signals, and coredumps).  Status between 128 and 192 (128+64) indicates the process died from a signal, so to mask the signal indication:  `141 - 128 = 13`.

`13` is `SIGPIPE` (see `kill -l`) so this means a process was killed when it was trying to write into a pipe that has been already closed.

Basically some process (vw?) failed and decided to exit and so another process that was trying to provide it with input got a `SIGPIPE` and died as well.

@kemaswill it would be very useful if you could provide more info:

- How was the spanning_tree process started when it failed?
- What was the correct way to start it?

Then perhaps someone can add more sanity checks to `spanning_tree` to avoid similar mistakes in the future.

**Comment by kemaswill** [comment]
@arielf Never mind, actually the spanning tree just fails to start due to permission issue, sorry for the false alert.
